### PR TITLE
Timed events

### PR DESCRIPTION
**Pull request by geom3trik**

### CODE REVIEW

**Comment by geom3trik** [comment]
Supersedes #24

File: crates/vizia_core/src/context/backend.rs [hunk 1]
```diff
@@ -264,4 +265,19 @@ impl<'a> BackendContext<'a> {
     pub fn needs_refresh(&mut self) {
         self.0.style.system_flags = SystemFlags::all();
     }
+
+    pub fn emit_scheduled_events(&mut self) {
```
Comment (DasLixou) [inline crates/vizia_core/src/context/backend.rs:269]: Shouldnt this be called like "wait_for_scheduled_events"?

**Comment by DasLixou** [comment]
So when I understand this correctly we are constantly popping and pushing in the BinaryHeap until a finished scheduled event is placed at the top, then we get this and execute it, right? Wouldn't it be smarter to just sort them from their time once and maybe use something like wakers in the async world instead of constantly whiling over it?

**Comment by geom3trik** [comment]
> So when I understand this correctly we are constantly popping and pushing in the BinaryHeap until a finished scheduled event is placed at the top, then we get this and execute it, right? Wouldn't it be smarter to just sort them from their time once and maybe use something like wakers in the async world instead of constantly whiling over it?

No that's not what's happening. A binary heap is already sorted, so each cycle it checks which scheduled events are ready to be sent and then pushes those events into the event queue by popping them off of the binary heap. These are two different structures, `event_queue` and `event_schedule`.

File: crates/vizia_core/src/context/backend.rs (hunk 1 above)
Comment (geom3trik) [reply crates/vizia_core/src/context/backend.rs:269]: Sure but you're also emitting those events by moving them to the event queue.

**Pull request merged**